### ISSUE TITLE
2.0 Discussion

### ISSUE THREAD

**Issue by timarney** [timarney/react-app-rewired#162]
TLDR:
See Readme notes for alternatives to this repo:

Also see - https://github.com/timarney/react-app-rewired/issues/343 as of Create React App 2.1.2 (there's a breaking issue) versions prior not using the utility functions from this repo still worked.

<hr>

Opening this to discuss work to be done in a 2.0 release.

One thing I would like thoughts on dropping support for older versions of Webpack.

i.e. this part of code
```
// Older versions of webpack have `plugins` on `loader.query` instead of `loader.options`.
  const options = loader.options || loader.query;
  options.plugins =  [pluginName].concat(options.plugins || []);
  return config;
```

**Comment by Guria** [comment]
We should invest some time into tests. But it isn't blocks major release I suppose

**Comment by timarney** [comment]
Would love to get some tests in place.  If you have any proposal as far as best setup for that would be  +1

**Comment by Guria** [comment]
We also can drop passing process.env.NODE_ENV as argument to rewires.

**Comment by timarney** [comment]
As per my note here
https://github.com/mobxjs/mobx-react/issues/381#issuecomment-354103843

should make some adjustments for plugin ordering i.e. https://github.com/loganfsmyth/babel-plugin-transform-decorators-legacy#note-order-of-plugins-matters

**Comment by andriijas** [comment]
Okey lets get the party started, i have a pending PR, 

@timarney can you create a `next` branch of react-app-rewired  and we can accumulate changes there?

**Comment by timarney** [comment]
There's a branch now.

**Comment by andriijas** [comment]
[x] Update less rewire to support new CSS modules in react-scripts 2.0
[ ] Deprecate support for old versions of react-scripts as mentioned above
[ ] Update rewires if needed

When CRA2 is out:

[ ] Update README to make it clear that react-app-rewired 2.0 only works with react-scripts 2.0
[ ] Update LESS readme to point to CSS module documentation in CRA

**Comment by Guria** [comment]
Is it critical to drop react-scripts@1.x support?
Do we expect critical config changes on CRA side that hard to manage on rewired side?

**Comment by andriijas** [comment]
Keeping 1.0 support will require more code comments to keep branching/if statements documented (or it will be easy to loose track.) but should be fine, not a lot of work to keep support. 

some stuff is old webpack legacy I think it can be removed safely.   Also webpack 4 is in the making, not sure if the that will add alterations on config compability  but hopefully the utilities rewire provides will just continue to work with minimal updates. 

On the other hand you could stay with current version of rewire for cra1? I mean if you stay with it it doesn’t make sense to upgrade to rewire either.

**Comment by pauliusuza** [comment]
Would it be possible for you to publish a test version of react-app-rewired under @next in npm?

**Comment by andriijas** [comment]
@timarney  now that there are changes that breaks rewire with cra2, can we have a pre release? I just pushed a minor tweak to next branch.

**Comment by timarney** [comment]
Should be there now:

next: 2.0.0
➜  react-app-rewired git:(next)

**Comment by andriijas** [comment]
Thanks. 

I figured its better to hold of cleaning of supporting old webpack config formats untill webpack 4 is in cra2.

**Comment by unconfident** [comment]
After thinking about #204 for a while I think I'd like to see better utility for rule/loader list traversal and manipulation. What I'd like to be able to out of the box with utils provided by rewire v2 or earlier is:

- Improved search logic that resembles Webpack's logic more closely.  
  Basically #204, will try to submit PR sooner or later
- Being able to find every rule that includes specific loader instead of just one
- Proper destinction between a [`Rule`] and [`UseEntry`], searching `Rule` by loader name or use entry 
- Inserting loader before or after specific rule
  - With an automatic conversion of a shorthand to list style. I.e. converting 
    ```js
    {
      test: /\.jsx$/,
      loader: 'babel-loader',
      options: { ... } 
    }
    ```
    to

    ```js
    {
      test: /\.jsx$/,
      use: [
        { loader: 'babel-loader', options: { ... } }
      ]
    }
    ```
- Changing loader options that would transparently convert string shorthand to a [`UseEntry`] object. I.e.

    ```js
    { test: /\.jsx$/,
      use: [
        'babel-loader'
      ] }
    ```
    to the code block above

[`Rule`]: https://webpack.js.org/configuration/module/#rule
[`UseEntry`]: https://webpack.js.org/configuration/module/#useentry

**Comment by timarney** [comment]
Noting this for discussion 

> What I'd like to propose is adding another function to the main package that would simillarly to injectBabelPlugin locate rules using babel-loader and switch babelrc option to true.
> This way it will be both, explicit and backwards compatible

https://github.com/timarney/react-app-rewired/issues/203

**Comment by nfantone** [comment]
@timarney Hi! Today I tried using `2.0.0` (+ latest `react-scripts@2.0.0-next`) and while the `build` command worked with no noticeable issues, `test` fails complaining about `babel` `6.26.0` usage.

```sh
    Requires Babel "^7.0.0-0", but was loaded with "6.26.0". If you are sure you have a compatible version of @babel/core, it is likely that something in your b
uild process is loading the wrong version. Inspect the stack trace of this error to look for the first entry that doesn't mention "@babel/core" or "babel-core"
to see what is calling Babel.
```

Using regular `react-scripts test` works as expected. Any advice on this?

**Comment by timarney** [comment]
@andriijas might be able to shed some light but there really hasn't been any work done on the next version for a while and it's on the back burner for me for now.

**Comment by unconfident** [comment]
By the way, I made that thing I suggested in [my previous comment](#issuecomment-366865659) into a separate package of its own: [`@unconfident/webpack-module-rules-utils`](https://www.npmjs.com/package/@unconfident/webpack-module-rules-utils)

I think it can be useful outside of react-app-rewired too. For instance Next.js' webpack configuration is also meant to be customized in a similar way at runtime. Another one I can recall would be Storybook's "[full control mode](https://storybook.js.org/configurations/custom-webpack-config/#full-control-mode)" and there are likely some other projects out there that suggest manipulating Webpack configuration at runtime

It has not yet been covered with tests and also lacking documentation at the moment. So if you're going to use it, use it with caution, and either rely on intellisense of your code editor or take a look at the [source code](https://github.com/unconfident/webpack-module-rules-utils/blob/cc4708ab219a5a47a7f66aeb678a2f5e762721a5/index.js#L186-L350) for the list of method names and their signatures.

**Comment by timarney** [comment]
Note: If anyone has interest in leading / maintaining  a 2.0 release @ me .

My time has been and will be limited for this project.  **That is to say considering not doing 2.0 given I can't support it** .  Thanks to all for helping out to date.

**Comment by andriijas** [comment]
Thx for the update. Im currently not involved in any react project therefor not putting any time on the tools I worked on last spring.

**Comment by AndyOGo** [comment]
what a bummer, this project is a great idea

**Comment by AndyOGo** [comment]
So I ask to incorporate this concept of rewiring into the official repo 😃 
https://github.com/facebook/create-react-app/issues/4970

@timarney 
@andriijas 
@kbrown
Please upvote if your like it and contribute to the feature request

**Comment by Timer** [comment]
Hey @AndyOGo!
We will not be considering adding a feature like this to Create React App, especially now with the release of Babel 7 and the more-often major versions of Webpack -- the possibility for churn is just too high.

However, I'd love to hear what use cases people have for "rewiring" configuration and add official support for said changes, e.g. TypeScript, Sass (already in 2.0), Vendoring (already in 2.0), Import SVG as React Component (already in 2.0), etc.

**Comment by AndyOGo** [comment]
@Timer 
Thanks for your super quick feedback.

Great you are watching this.
And thanks for the news about `2.0`.

Personally I hope that all those great ideas could be accumulated into one maintained repo.

**Comment by dawnmist** [comment]
@Timer 

We've used react-app-rewired fairly extensively to tweak the configuration without having to lose all future updates/improvements in the build process, so I guess a list of what we use it for would be useful. :)

For the build/dev build config:
* Less/Sass support.
* Modify the output filenames for a specific group of images to put them into a subdirectory under /static/media/twemoji (Twemoji svgs).
* Typescript customisation (fork-ts-checker-webpack-plugin, cache-loader, thread-loader). We already use create-react-app-typescript as our base, but there are a few bits of the build that needed tweaking in order to reduce the build time to something more reasonable.
* CommonsChunkPlugin (though that will need replacing with the new splitting when moving to Webpack 4) to pull out common vendor code/common chunk code from code-split sections of the app.
* Build profiling (config.profile = true) to find out where the time in the build was going.
* BundleAnalyzerPlugin if requested (to see what is being included in bundles that shouldn't be - e.g. pulling in the whole of lodash when intending to use only a single function, or to find items that are included in the wrong bundle/included multiple times, etc). It's able to be enabled or disabled by an environment variable.
* EnvironmentPlugin - to inject git version numbers/sha/etc to the build component, so we know what version has been deployed to the test server/etc.
* To stop the dev build from creating all bundles as "bundle.js" - production works fine with this, but dev was overwriting code split/commons chunk bundles by trying to call all bundle files `static/js/bundle.js`.

For the WebpackDevServer settings:
* Setting HTTPS certificate variables for WebpackDevServer.

For Jest:
* Customising config.transform for typescript instead of javascript.
* Setting config.transformIgnorePatterns so that a couple of directories inside node_modules get processed (internal code projects included as modules)
* Setting config.moduleNameMapper so that css classnames for css modules get written out as the local key (i.e. :local(.myClassName) is written as "myClassName" instead of as blank), so that we can test that items which use different styles depending on various conditions are actually using the right styles/executing the right code paths.
* Turning on useBabelrc - with a code-split app, we need to use the "dynamic-import-node" babel plugin during testing in order to be able to test some parts of the app (as the node environment it runs in doesn't have support for dynamic imports by default).
* Setting config.collectCoverageFrom to limit coverage reporting to areas that actually have tests.
* Setting config.testMatch - there's a helper file in some test directories that is used for reused test functions so that the same functions don't need to be rewritten in each specific test file, but we then need to exclude that file from being run as a test itself.
* Adding the enzyme-to-json/serializer to config.snapshotSerializers for better diff/readability of snapshot tests, so that people can see differences more easily when checking whether the change is reasonable compared to the code they have just changed.

Some of that is probably worth adding to create-react-app like the jest config for css classnames so that classnames are actually included in snapshot data and can be used to test that the correct code paths are being run given the input data given to a component, or having the ability to _optionally_ generate bundle statistics in the build process. Other parts are fairly specific to our project, but are in reality still tweaks of the default config and are not items we really wanted to lose future updates from create-react-app/create-react-app-typescript in order to configure.

One of the most common uses I've seen reported in issues is to be able to inject various babel plugins in order to support use of things such as relay, react-cosmos, antd, mobx, support for decorators, etc.

**Comment by biokillos** [comment]
@Timer thanks heaps! Personally, we have it only to support `Ant Design`. Please upvote for such a use case.

**Comment by timarney** [comment]
I haven't used MDX yet but that have a CRA example using this repo.

**MDX + Create React App**
https://github.com/mdx-js/mdx/tree/master/examples/create-react-app

**More stats**
https://libraries.io/npm/react-app-rewired/usage?page=1

**Comment by okwolf** [comment]
We're nearing the 2.0 release: https://github.com/facebook/create-react-app/issues/5103

**Comment by arackaf** [comment]
Hi everyone. I've read through this thread, and just curious where this project stands. It seems Tim is busy on other things, which is cool.

Have other people (graciously) decided to come in and maintain this? I'm just trying to figure out if I should hold out for an update supporting CRA 2, or if I should just part ways with CRA.  Being a MobX user I kind of dependent on this project.

Anyway - this project is incredible - no complaints. Just asking what the future looks like here.

**Comment by timarney** [comment]
If anyone is interested this file contains the basics of how things work.  Nothing special.

https://github.com/timarney/react-app-rewired/blob/master/packages/react-app-rewired/scripts/start.js

```
// load original configs
const webpackConfig = require(webpackConfigPath);
const devServerConfig = require(devServerConfigPath);
// override config in memory
require.cache[require.resolve(webpackConfigPath)].exports =
  overrides.webpack(webpackConfig, process.env.NODE_ENV);

require.cache[require.resolve(devServerConfigPath)].exports =
  overrides.devServer(devServerConfig, process.env.NODE_ENV);

// run original script
require(paths.scriptVersion + "/scripts/start");
```

**Comment by okwolf** [comment]
I've used this project as inspiration for implementing [something similar for Hyperapp](https://github.com/okwolf/cra-hyperapp/blob/master/bin/index.js). Perhaps I will learn from updating that project for `react-scripts@2.0`.

**Comment by cj** [comment]
@Timer The reason I use react-app-rewired for "rewiring" the configuration, is to add an alias '~' that points to my src dir which avoid paths like `../../../` and to add react-hot-loader.

The current version of `react-app-rewired@1.6.2` works perfectly with create-react-app v2 and `react-scripts@2.0.1`, here's my `config-overrides.js`:

```javascript
/* eslint-disable no-param-reassign */
const { resolve } = require('path')

module.exports = {
  webpack(config) {
    const srcDir = resolve('./src')
    const babelLoader = config.module.rules
      .find(rule => Object.keys(rule).includes('oneOf')).oneOf
      .find(rule => Object.values(rule).includes(srcDir)).use
      .find(rule => rule.loader.includes('babel-loader'))

    babelLoader.options.plugins.push('react-hot-loader/babel')

    config.resolve.alias['~'] = resolve('./src')

    return config
  },

  jest(config) {
    config.moduleNameMapper['~(.*)$'] = '<rootDir>/src/$1'

    return config
  },
}
```

**Comment by cj** [comment]
@fooey the above code I pasted works with CRA 2.0, so you can still continue to use rewired for hot reloading until it's officially supported!

**Comment by arackaf** [comment]
Pretty much everything will be different with CRA 2, given the deep changes webpack underwent. I’m inclined to think a separate project might be simpler than trying to keep two separate versions of CRA support straight, either with version numbers or separate project exports. But if Tim wants to keep this going I’d be happy to help integrate my code over there.

**Comment by timarney** [comment]
Ultimately I think moving over to a fresh code base with no legacy CRA versions makes sense.

Biggest thing outside of time is, I'm not using CRA these days so I'm not in the game. That's an important piece in my eyes as far as maintaining.

Question is best way to redirect users if they want to move forward with "the hack".

Not sure on the wording yet but rough todos:

1. Add a note to the npm install process to mention - "Don't use this for CRA 2"  As @arackaf notes https://twitter.com/AdamRackis/status/1045805923988647936 technically it does but I would rather get people moved over.

2. Can link out on the homepage of this repo to https://github.com/arackaf/customize-cra

3. Continue to mention heavily that you're breaking the "guarantees" that CRA provides.  Use at your own risk etc... have you tried  Razzle or Next.js https://github.com/zeit/next.js/

> "Stuff can break" — Dan Abramov
https://twitter.com/dan_abramov/status/1045809734069170176

4. I'll also note on this post - https://medium.com/@timarney/but-i-dont-wanna-eject-3e3da5826e39

<hr>

**Comment by arackaf** [comment]
Sounds good to me.  I'm using CRA 2 at work, so keeping customize-cra up to date will definitely be something I make time for, by necessity.  

I link in the readme here, directing people there would likely help. And yeah, I've got Dan's quote in there already :)

PRs are absolutely open, so if anyone in this community wants to convert any of the helpers here, please don't be shy.

**Comment by penx** [comment]
> However, I'd love to hear what use cases people have for "rewiring" configuration and add official support for said changes

@timarney eslint configuration is pretty much the only thing I (and other React developers I know) continue to eject/rewire/avoid CRA for

**Comment by cleverettgirl** [comment]
@Timer monorepo/yarn workspaces support -- specifically sharing jsx components outside of a CRA's src folder with multiple CRA's. I know this was dropped in v2 and replaced with support for nwb, however i'm still waiting on that excellent documentation ;). The only reason we didn't go straight to integrating with nwb was because it seemed a little more than what we needed as we're not interested in publishing any packages to npm.

**Comment by jgoux** [comment]
Hello all,

I talked a bit about my vision of the future of react-app-rewired [here](https://github.com/arackaf/customize-cra/issues/2).

@arackaf would you be interested in registering the `@react-app-rewired` organization if @timarney is ok with it?

I think it would be great to house all the "plugins" under the same organization, and if people want their own presets, like you did @arackaf , then they can just make their own repository and compose what they want.

**Comment by arackaf** [comment]
Sounds good to me. Just let me know what I’d need to do. I’ve never registered under a shared org before with npm.

**Comment by timarney** [comment]
> @arackaf would you be interested in registering the @react-app-rewired organization if @timarney is ok with it?

Happy with whatever makes sense for the community behind this.

**Comment by arackaf** [comment]
@timarney - thanks for being so cool :)

Forgive my ignorance, this is a bit new to me - how do I go about registering that organization, and publishing customize-cra under it?  I assume the code would now be released under 

`@react-app-rewired/customize-cra`

or whatever?

**Comment by timarney** [comment]
@arackaf 
> how do I go about registering that organization, and publishing customize-cra under it? I assume the code would now be released under

Uncharted waters for myself as well.  Not sure how that works.

**Comment by arackaf** [comment]
Hm - I won't sweat it too much then. Seems like just a semantic nice-to-have.

What I most curious about, if you have another moment, is the reasoning / purpose behind the `env` variable. Ditching that would simplify some things, but we don't wanna break anything 

https://github.com/arackaf/customize-cra/issues/2#issuecomment-426668112

**Comment by jonathantneal** [comment]
Hey, I’ve setup orgs before on GitHub and npm. I did so for csstools:

- https://www.npmjs.com/org/csstools
- https://github.com/csstools/

I’m happy to jump on a call if you want a walk-through.

Publishing public projects on scoped npm is one command trickier than normal, but nothing too weird.

```bash
npm publish --access=public
```

**Comment by rjerue** [comment]
> @Timer monorepo/yarn workspaces support -- specifically sharing jsx components outside of a CRA's src folder with multiple CRA's. I know this was dropped in v2 and replaced with official support for nwb, however i'm still waiting on that excellent documentation ;). The only reason we didn't go straight to integrating with nwb was because it seemed a little more than what we needed as we're not interested in publishing any packages to npm.

This is also my exact use case.

I've been using CRA for by build process in a project that's a monorepo with a react native project and would love to upgrade to CRA2, but I was using this to help with the monorepo in development.

**Comment by cleverettgirl** [comment]
@amcvitty yes! We were using yarn workspaces, which I agree seemed like a simple and great solution. (This isn't to say I don't understand why they dropped, just bummed.) We also have a very similar setup, but our "core" does have some jsx. I hadn't heard of yalc, but it seems like it could be promising-- thank you for posting that link and sorry I cant offer any help. We're kinda researching all our different options while we wait for nwb documentation to see what will be the best for us. Right now it seems like we might be leaning towards forking cra2 and trying to modify react-scripts.

**Comment by amcvitty** [comment]
@cleverettgirl @rjerue I did some more work on this and came up with something that works. I still don't know if it's a great idea, but I think it's good enough for our use case. 

Basically we've reintroduced react-app-rewired (v1.6.2) into the build to allow us to include "core" in the modules that run through babel in webpack. We've taken out the yarn workspaces config, but we've kept the same hierarchy on the filesystem. 

So - we have 4 package.json files in total. 

ROOT/package.json (only build tools here, as devDependencies. Includes yalc )
ROOT/core/package.json (simple JS module - no dependencies. Doesn't do any babel transpile, publishes ES2018 source files with JSX)
ROOT/web/package.json  (from CRA with react-scripts: 2.0.3)
```
 dependencies: { 
    ... 
    "core": "file:.yalc/core",
    ... 
}
```
ROOT/native/package.json  (from CRNA, also depends on core with `yalc link`)


Our build steps from CircleCI look like: 
```
 - run: yarn install
 - run: cd core && yarn install && yarn lint:all && yarn flow && yarn test  && ../node_modules/.bin/yalc publish
 - run: cd web    && ../node_modules/.bin/yalc add  core && yarn install && yarn lint && yarn test
 - run: cd native && ../node_modules/.bin/yalc link core && yarn install && yarn lint && yarn test
```
(I'm not sure why yalc link works better than yalc add from core. YMMV.)

Then in web/config-overrides.js we include code in the babel-loader like this. (this is probably messier than it needs to be). 
```
/* eslint-disable no-param-reassign */

module.exports = {
  webpack(config) {
    // Explicitly include core in the babel transform
    const babel = config.module.rules
      .find(rule => "oneOf" in rule)
      .oneOf.find(rule => /babel-loader/.test(rule.loader));

    if (!Array.isArray(babel.include)) {
      babel.include = [babel.include];
    }
    babel.include = babel.include.concat(/node_modules\/core\//);

    return config;
  },

  jest(config) {
    // Explicitly include core in the babel transform
    config.transformIgnorePatterns[0] =
      "[/\\\\]node_modules[/\\\\]!(core/).+\\.(js|jsx)$";

    return config;
  },
};
```

All of this means we can have webpack, jest and flow all pointing directly at the original source files in core; the imports can say `import 'core/component/MyComponent.js'` instead of `import 'core/lib/component/MyComponent'`. Importing the transpiled source works fine for webpack, but not for flow, which needs the original files. Also, source maps for web now also include the core stuff, which is really good.

So - at the cost of some hacky files and a couple of extra build steps, we've got full tooling (jest/flow/eslint) on CRA and CRNA in a monorepo that's all versioned together.

**Comment by rjerue** [comment]
@amcvitty just gave it a try in a test project, seems to be working

I do though need the JSX to transpile as well. I'd actually expect it to as those files should be covered by the babel loader though yeah?

Maybe when I have time, adding a watch to the my common dir too to have it run yalc publish and yalc update.

**Comment by cleverettgirl** [comment]
@amcvitty this is working with CRA2 as well?

**Comment by jonathantneal** [comment]
I was able to update https://github.com/csstools/react-app-rewire-postcss/ to work with CRA2, both in development and production builds.

**Comment by marcopeg** [comment]
Hi guys, I'm working on a [`create-react-app` 2.x fork](https://github.com/marcopeg/create-react-app/blob/master/packages/react-scripts/README.md) that aims to bring back compatibility with `react-app-rewired` API.

So far I have plain support for the extension API of both webpack and webpackDevServer config, plus the support for custom app templates.

I would love to gain some collaborators that can help me maintain and to extend this :-)

**Comment by amcvitty** [comment]
@cleverettgirl @rjerue yes this is working with JSX and CRA2. We've had this in master for about a week. The only issues so far are that you have to go run things in each package instead of a single yarn install, and changes in core won't get reflected in web without restarting `yarn start`. so - not as convenient, but it's pretty solid

**Comment by arackaf** [comment]
@marcopeg fyi I've got a set of 2.0-compat plugins working for CRA 2.0 here: https://github.com/arackaf/customize-cra

**Comment by Timer** [comment]
TypeScript is now officially supported as of Create React App 2.1. [Read the release notes to get started](https://github.com/facebook/create-react-app/releases/tag/v2.1.0)!

We're continuing work on supporting more use-cases and will reinstate support for monorepos (in some form) and a modern build mode soon!
Still keeping tabs on this issue.

**Comment by PinkaminaDianePie** [comment]
@Timer no matter how many stuff would be supported by CRA out of the box, it won't solve this issue. You just can't predict all of the use cases and you can't add everything to CRA. That's why we need this tool. Our use case is not "we need TS support" or "we need SCSS", our case is "we need to configure CRA any way we want, but without ejecting".

**Comment by Timer** [comment]
> our case is "we need to configure CRA any way we want, but without ejecting"

That's OK, especially for advanced users. What's concerning for us is that many beginners use this and will have no idea what to do when it breaks (because they C+P'd it from a blog post). :-)

My objective is not to make this tool obsolete, but to reduce reliance on it. I think this tool is awesome for people who know webpack.

**Comment by PinkaminaDianePie** [comment]
> What's concerning for us is that many beginners use this and will have no idea what to do when it breaks
Well, 2.0 covered some of my cases (babel 7 and SCSS) but not entirely. I still need some custom loaders, which won't be included in official CRA package.

Best way to help the community - provide some official hooks, which people will be able to use instead of this tool. You can't forbid people to configure react scripts. People will do it anyway. But for now they rely on 3rd party tools and blog posts of people, who don't even have time to support this tools. Why not to give some unsafe but official way? If people will struggle with broken 3rd party tools and packages, they will blame not their authors, but CRA team, because instead of using some official way they forced to use 3rd party tools. You can say that official way is ejecting but 3.6k stars on this package telling me that it's not a way at all. Supporting entire webpack config after ejecting with bunch of helpers, like error overlay is a magnitude harder than to support one hook with react-app-rewired.

**Comment by Timer** [comment]
Official hooks are off the table because that'd imply a safety that isn't really there. Things **will** break.

Most use cases we find people overriding configuration for can actually be easily achieved without touching configuration. Giving people an "official" way will just make this more prolific.

**Comment by dawnmist** [comment]
@Timer I've been trying to simplify our build to the minimal required (e.g. changing how git version numbers are passed into the application so that we didn't need to use webpack.EnvironmentPlugin to add them at compile time). Out of that, I found I've missed a few:
* Ability to _append_ an external/non-bundled css file. Our app is the front-end to a suite of applications used by multiple clients. The end client needs to be able to provide a corporate colors theme css file that is not bundled into the downloadable build, and to be able to extend the content of that css file to match changes in their database's contents (e.g. contrived but equivalent example: setting a color to use for a post type when they add a new post type to the DB). The cra bundled css is added to the very end of the head tag in the html template file, which means that there is no way I can find to be able to link in a specific externally-editable css file as the _last_ css file in the list. I did try using the html-webpack-plugin's template directives in our html file to specify to add the extracted css before the manual link, but that still gets overridden such that those css files get added twice - once where I requested they be added, then again at the very end of the head tag where react-scripts(-ts) adds them. If there is a way to achieve this without using `html-webpack-include-assets-plugin` I'd love to hear it.
* Customizing `fileLoader.options.name`, so that svg files from `twemoji` were put into `static/media/twemoji/` instead of `static/media/` (with the intent that the use of a subdirectory could be extended to include other major graphics providers such as svgs for icons, etc, at a later date).

I also realised (and updated the original list) that we also store our storybook files in the `__tests__` directories, so Jest `config.testMatch` is required to be able to exclude `*.stories.ts(x)?` as well as the `__helpers__.ts(x)?` file. Those use the storyshots addon to turn the stories into snapshot tests (and in order for them to actually be usable snapshot tests the css class names need to be output to the snapshot - which is how that all tied together).

I admit that there is quite a lot that we have been customizing - but most are small tweaks to enhance/improve/add missing functionality to what create-react-app provides by default. Other than during major version upgrades using react-app-rewired rather than ejecting has meant that when there have been improvements in the build tools used in create-react-app, we've been able to "just receive" them. The ability to receive upstream updates to the build tools is absolutely worth doing all we can to stay attached - we simply don't have the amount of time required to maintain the entire set ourselves properly, so being able to simply maintain our tweaks instead is extremely valuable.

**Comment by cj** [comment]
FYI if anyone is interested, I found this the other day https://github.com/sharegate/craco and it works great!

**Comment by ndbroadbent** [comment]
Thanks so much @cj, I would have been totally stuck if you didn't share craco! It's also working great for me, and it was really easy to set up Ant Design, Less, and Preact.

Here are some example `craco.config.js` configuration files:

* [Less](https://github.com/sharegate/craco/blob/00ae1a4364bfbe5b1ef7d5f5d89affc030c46c1f/recipes/use-less-loader/craco.config.js)
* [Ant Design + Less + modifyVars](https://github.com/sharegate/craco/blob/0c35cc366ea79250e31f161172014b6a7107e3d3/recipes/use-ant-design/craco.config.js)
* [Ant Design + Less + modifyVars + Preact](https://gist.github.com/ndbroadbent/f7d6a343c5f4976ef4f8fa10a8595bad)
* [Preact](https://github.com/sharegate/craco/blob/d503f663bf8cfa9e222d37f180b49ae4c2d082ca/recipes/use-preact/craco.config.js)

There are [some more examples in the `/recipes` directory in the craco repo](https://github.com/sharegate/craco/tree/master/recipes).

I hope that helps someone! (It would have really helped me when I was just getting started.)

**Comment by abenhamdine** [comment]
> Here are some example craco.config.js configuration files:

> Less
> Ant Design + Less + modifyVars

It's worth noting that these configurations still works with CRA2.1.1 and react-app-rewired, we continue to use it without no problem in our application.

Our config-override.js for anyone interested : 
```js
const { injectBabelPlugin } = require('react-app-rewired')
const rewireLess = require('react-app-rewire-less')
const { BundleAnalyzerPlugin } = require('webpack-bundle-analyzer')
const rewireTsJest = require('react-app-rewire-ts-jest')
const WebpackBar = require('webpackbar')

module.exports = {
	webpack: function (config, env) {

		const path = require('path')
		// For import with absolute path
		config.resolve.modules = [path.resolve('src')].concat(config.resolve.modules)

		config = injectBabelPlugin(['import', { libraryName: 'antd', style: true }], config)  // change importing css to less

		const options = {
			javascriptEnabled: true,
			// We use modifyVars option of less-loader here, you can see a orange button rendered on the page after reboot start server.
			modifyVars: {
				"@primary-color": "#f48b31",
				"@font-size-base": "16px",
				"@font-family": "Nunito",
				"@layout-header-background": "#fff",
				"@layout-trigger-color": "#565558",
				"@layout-trigger-background": "#fff"
			},
		}
		config = rewireLess.withLoaderOptions(options)(config, env)

		/*
		=== webpack plugins ===
		Note : we prefer to add new webpack plugins here by ourselves as following, rather than using third-party helpers like rewire-some-plugin-XXX
		because they add an unecessary abstraction only to add a plugin in config.plugins array
		*/
		config.plugins = (config.plugins || []).concat([
			new BundleAnalyzerPlugin(),
			new WebpackBar({
				profile: true,
			})
		])

		return config

	},
	jest: function (config) {

		// placeholder
		const configTsJest = {}
		config = rewireTsJest(config, configTsJest)
		config = {
			...config,
			reporters: [
				"default",
				["./node_modules/jest-html-reporter", {
					pageTitle: "Payroll-app-next: Tests suite",
					outputPath: "tests/report/index.html",
					includeFailureMsg: true,
					dateFormat: "dd/mm/yyyy HH:MM",
					sort: 'status'
				}]
			],
			collectCoverage: true,
			collectCoverageFrom: [
				"src/**/*.{ts,tsx}"
			],
			coveragePathIgnorePatterns: [
				"src/state/state.d.ts"
			],
			coverageReporters: [
				"html",
				"json"
			]
		}

		return config
	}
}
```

**Comment by ndbroadbent** [comment]
> It's worth noting that these configurations still works with CRA2.1.1 and react-app-rewired, we continue to use it without no problem in our application.

OHHHH. Ahaha well I guess that was a waste of time. The README says "This repo doesn't support CRA 2.0", so I kind of assumed that it was broken.

GitHub also didn't help by collapsing ~30 comments in the middle, so I didn't see anything about [customize-cra](https://github.com/arackaf/customize-cra). For some reason I thought GitHub only collapsed the "+1" and "me too" comments.

I might just use your config then. Thanks!

**Comment by ndbroadbent** [comment]
Sorry I just wanted to follow-up one last time and say that I actually really like [craco](https://github.com/sharegate/craco) to configure `create-react-app`. The architecture is awesome, and it's very easy to install and get it running. I also decided to release a few plugins as NPM packages:

* [craco-antd](https://www.npmjs.com/package/craco-antd) - For Ant Design, including Less and `babel-plugin-import`
* [craco-less](https://www.npmjs.com/package/craco-less) - Adds `less-loader` to the webpack config

I think my `craco.config.js` file is super clean and just does exactly what I need:

```js
const CracoAntDesignPlugin = require('craco-antd');
const { BundleAnalyzerPlugin } = require('webpack-bundle-analyzer');
const WebpackBar = require('webpackbar');

module.exports = {
  webpack: {
    alias: { react: 'preact-compat', 'react-dom': 'preact-compat' },
    plugins: [
      new BundleAnalyzerPlugin(),
      new WebpackBar({ profile: true }),
    ],
  },
  plugins: [{ plugin: CracoAntDesignPlugin }],
};
```

Thanks @abenhamdine for the tip about `BundleAnalyzerPlugin` and `WebpackBar`, those are awesome!

<img width="626" alt="screen shot 2018-11-13 at 5 33 33 am" src="https://user-images.githubusercontent.com/139536/48379410-a5116200-e706-11e8-9c4e-7dd3efc10a79.png">

**Comment by abenhamdine** [comment]
Sorry for the noise, I just want to let know that craco and craco-antd plugin work perfectly with cra 2.1.1/antd/typescript so craco looks like a good alternative.  

In another hand, react-app-rewired is also still working fine for us with cra 2.1.1/antd/typescript (maybe some react-app-rewired-xxx plugins are breaking with cra 2.X but it didn't affect us).

**Comment by AndyOGo** [comment]
@kopax 
Just try https://github.com/sharegate/craco as first mentioned above https://github.com/timarney/react-app-rewired/issues/162#issuecomment-436120752

**Comment by kopax** [comment]
Well, my issue was a wrong parameter of `fs.inotify`. Solved. And so far I am using react-app-rewired only for plugin babel-plugin-styled-components, it seems to work fine with CRA v2. I've also seen a few people saying it work fine here, so I presume craco is not required.

**Comment by dawnmist** [comment]
The core functions work with CRA2.x, but many (most?) of the pre-existing rewires _do not_, because they were written for Webpack 2/3, and the Webpack 4 configuration has altered significantly - many of those rewires break. The decision finally was that there should be a split to a new project for CRA2.x rather than continuing react-app-rewired so that old rewires that were no longer maintained and were not CRA2.x compatible were not mistakenly thought to be working.

Two potential splits have emerged - one being craco and the other being customize-cra. Customize-cra uses the core functionality from react-app-rewired, but provides new rewires written for CRA2.x. Craco provides that functionality in a separate way without requiring react-app-rewired.

**Comment by comerc** [comment]
@Timer just add API-hook, like GatsbyJS [onCreateWebpackConfig](https://www.gatsbyjs.org/docs/node-apis/#onCreateWebpackConfig). Please! I like CRA, but GatsbyJS leading.

**Comment by jbergens** [comment]
The things I've used rewired for is mostly getting decorators to work. It would also be nice to get to override some other things like eslint config.

**Comment by jamesmfriedman** [comment]
For anyone just manually overriding the config, this works 100% fine with CRA2, using it in production with many projects.

**Comment by jihchi** [comment]
@Gamezpedia #343

**Comment by AndyOGo** [comment]
@jihchi 
@jamesmfriedman 
@Gamezpedia 
Just switch to CRACO JS (crafted for CRA V2)
https://github.com/sharegate/craco

**Comment by abenhamdine** [comment]
> Just switch to CRACO JS (crafted for CRA V2)

It's not so simple, https://github.com/sharegate/craco is also broken with cra 2.1.1, see https://github.com/sharegate/craco/issues/61

**Comment by AndyOGo** [comment]
That's a bummer. At least it was written for CRA V2, so it should at least be more likely that it is maintained and will be fixed sooner (fingers crossed)...

Anyway I would create an issue at CRACO

**Comment by jamesmfriedman** [comment]
Again, no need to switch. This repo continues to work perfectly for CRA 2.1.1

**Comment by abenhamdine** [comment]
> Again, no need to switch. This repo continues to work perfectly for CRA 2.1.1

You're right, sorry, I read CRA 2.1.2 instead of 2.1.1, I edited my previous comment accordingly https://github.com/timarney/react-app-rewired/issues/162#issuecomment-449992045

This repo also still works perfectly for us until cra 2.1.1

**Comment by timarney** [comment]
I've published a 2.0 version to fix issues with CRA v2.1.2 .

Thanks @tiffon + others who.  Utility helper rewires have been removed.

If you must rewire this should get things back to a working state.  

Not other updates are planned this was just a break fix situation.

Also with that I'm going to close this issue.

**Comment by kopax** [comment]
Great, we decided to turn with craco that is now not yet updated to v2.1.2. Funny situation. Anyway thanks again @timarney we may swap back soon.

**Comment by AndyOGo** [comment]
seems rescripts is working:
https://github.com/rescripts/rescripts

**Comment by jamesmfriedman** [comment]
Thanks @timarney! Myself and other others I work with are very reliant on this project so I’d be happy to contribute to keep it going. 

I rely solely on custom requiring in config overrides, I’ve become a bit of a surgeon in regards to tweaking the CRA build process :). 

Love this project, thanks for your hard work.

**Comment by smmoosavi** [comment]
You can use [monkey-react-scripts](https://github.com/monkey-patches/monkey-react-scripts) 

It lets you mutate webpack config and change babel config for tests

```js
module.exports = function(webpackConfig, isDevelopment) {
  // mutate webpackConfig
};
```
